doc/rst/technotes/optimization.rst

```diff
@@ -0,0 +1,662 @@
+each locale to create one. That means that creating a slice, as in
+``MyDistributedArray[1..10]`` can actually be quite slow.
+
+One common scenario is that data needs to be copied between regions of
+arrays. For example
+``MyDistributedArray[1..10] = MyOtherArray[11..20];``.
+The compiler can optimize this kind of
+assignment in many cases today with Array View Elision (AVE). However, in
+order to optimize it in this way, both sides of the ``=`` need to be
+slice expressions. Note that in some cases, one might create a local
+array to store the contents of a remote slice; that can be done with AVE
+by redundantly slicing the local array, e.g.:
+``var Loc:[1..10] int; Loc[1..10] = MyDistributedArray[1..10];```
+
+Unoptimized Distributions
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The Chapel programming language is designed to support many distributions
+for domains and arrays. That includes distributions created by users.
+However, the present situation is that performance of Chapel programs
+depends on optimizations in the implementations of these domain/array
+distributions. These optimizations are present and reasonably well tuned
+for the Block distribution. Other distributions might not be optimized
+and have scaling issues.
+
+Performance Problems with Multidimensional Zippered Iteration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Issue: https://github.com/chapel-lang/chapel/issues/13147)
+
+Zippered iteration for multidimensional arrays/domains is much slower
+than zippered iteration for 1D arrays/domains. Since promoted calls, such
+as ``MyArray + MyOtherArray`` are implemented with zippered iteration,
+this problem also applies to that case.
+
+Potential ways to avoid this problem:
+ * express the computation with nested loops per dimension
+ * use 1D arrays and explicitly compute 1D indices from 2D conceptual indices
+ * create a 1D copy of the array (with ``reshape`` -- note that in the
+   future, we expect to have a way to ``reshape`` without copying)
+
+Cooperative Scheduling and Remote Tasks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Chapel's tasking model currently uses cooperative scheduling. That means
+that, once a task starts running on a core, it will be the only thing
+running on that core until the task either yields execution or ends.
+
+Occasionally, this can lead to problems when combined with ``on``
+statements. The ``on`` statements generally create a task on a remote
+locale. Those remote tasks will never get a chance to run if all of the
+cores are busy with existing tasks. This problem is rare, but it can
+cause performance issues if it comes up. The solution is to
+periodically call :ref:`currentTask.yieldExecution() <Yield_Task_Execution>`
+from any polling loops. This is already done in
+``waitFor()`` methods on atomics (see also
+:ref:`Functions_on_Atomic_Variables`).
+
+Tools for Understanding Performance
+-----------------------------------
+
```

Comment:
Consider another note here that says "This note is a draft and this section is TBD" (in prettier words than that 😄 )